src/sync/mutex.go

```diff
@@ -41,35 +40,132 @@ func (m *Mutex) Unlock() {
 }
 
 type RWMutex struct {
-	m       Mutex
-	readers uint32
+	// waitingWriters are all of the tasks waiting for write locks.
+	waitingWriters task.Stack
+
+	// waitingReaders are all of the tasks waiting for a read lock.
+	waitingReaders task.Stack
+
+	// state is the current state of the RWMutex.
+	// Iff the mutex is completely unlocked, it contains rwMutexStateUnlocked (aka 0).
+	// Iff the mutex is write-locked, it contains rwMutexStateWLocked.
+	// While the mutex is read-locked, it contains the current number of readers.
+	state uint32
 }
 
+const (
+	rwMutexStateUnlocked = uint32(0)
+	rwMutexStateWLocked  = ^uint32(0)
+	rwMutexMaxReaders    = rwMutexStateWLocked - 1
+)
+
 func (rw *RWMutex) Lock() {
-	rw.m.Lock()
+	if rw.state == 0 {
+		// The mutex is completely unlocked.
+		// Lock without waiting.
+		rw.state = rwMutexStateWLocked
+		return
+	}
+
+	// Wait for the lock to be released.
+	rw.waitingWriters.Push(task.Current())
+	task.Pause()
 }
 
 func (rw *RWMutex) Unlock() {
-	rw.m.Unlock()
+	switch rw.state {
+	case rwMutexStateWLocked:
+		// This is correct.
+
+	case rwMutexStateUnlocked:
+		// The mutex is already unlocked.
+		panic("sync: unlock of unlocked RWMutex")
+
+	default:
+		// The mutex is read-locked instead of write-locked.
+		panic("sync: write-unlock of read-locked RWMutex")
+	}
+
+	switch {
+	case rw.maybeUnblockReaders():
```

Comment:
Ended up reading the stdlib RWMutex implementation to see that they have the same priority of writer->readers over writer->writer lock.